src/backend/distributed/commands/table.c

```diff
@@ -593,59 +616,213 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand)
- * the foreign keys that given ALTER TABLE statement defines.
+ * CompareRangeVarsByOid is a comparison function to sort RangeVar object list.
+ */
+static int
+CompareRangeVarsByOid(const void *leftElement, const void *rightElement)
+{
+	RangeVar *leftRangeVar = *((RangeVar **) leftElement);
+	RangeVar *rightRangeVar = *((RangeVar **) rightElement);
+
+	/*
+	 * Any way we will check their existence, so it's okay to map non-existing
+	 * relations to InvalidOid when sorting.
+	 */
+	bool missingOk = true;
+
+	/*
+	 * As this is an object comparator function, there is no way to understand
+	 * proper lock mode. So assume caller already locked relations.
+	 */
+	LOCKMODE lockMode = NoLock;
+
+	Oid leftRelationId = RangeVarGetRelid(leftRangeVar, lockMode, missingOk);
+	Oid rightRelationId = RangeVarGetRelid(rightRangeVar, lockMode, missingOk);
+	return CompareOids(&leftRelationId, &rightRelationId);
+}
+
+
+/*
+ * GetAlterTableAddFKeyRightRelationIdList returns a list of oid's for right
+ * relations involved in foreign keys to be defined by given ALTER TABLE command.
+ */
+static List *
+GetAlterTableAddFKeyRightRelationIdList(AlterTableStmt *alterTableStatement)
+{
+	List *rightRelationRangeVarList =
+		GetAlterTableAddFKeyRightRelationRangeVarList(alterTableStatement);
+	List *commandList = alterTableStatement->cmds;
+	LOCKMODE lockMode = AlterTableGetLockLevel(commandList);
+	bool missingOk = alterTableStatement->missing_ok;
+	List *rightRelationIdList =
+		GetRelationIdListFromRangeVarList(rightRelationRangeVarList, lockMode, missingOk);
+	return rightRelationIdList;
+}
+
+
+/*
+ * GetAlterTableAddFKeyRightRelationRangeVarList returns a list of RangeVar
+ * objects for right relations involved in foreign keys to be defined by
+ * given ALTER TABLE command.
+ */
+static List *
+GetAlterTableAddFKeyRightRelationRangeVarList(AlterTableStmt *alterTableStatement)
+{
+	List *constraintList = GetAlterTableAddFKeyConstraintList(alterTableStatement);
+	List *rightRelationRangeVarList = GetRangeVarListFromConstraintList(constraintList);
+	return rightRelationRangeVarList;
+}
+
+
+/*
+ * GetAlterTableAddFKeyConstraintList returns a list of Constraint objects for
+ * foreign keys that given ALTER TABLE to be defined by given ALTER TABLE command.
  */
 static List *
-GetAlterTableStmtFKeyConstraintList(AlterTableStmt *alterTableStatement)
+GetAlterTableAddFKeyConstraintList(AlterTableStmt *alterTableStatement)
 {
-	List *alterTableFKeyConstraintList = NIL;
+	List *foreignKeyConstraintList = NIL;
 
 	List *commandList = alterTableStatement->cmds;
 	AlterTableCmd *command = NULL;
 	foreach_ptr(command, commandList)
 	{
-		List *commandFKeyConstraintList = GetAlterTableCommandFKeyConstraintList(command);
-		alterTableFKeyConstraintList = list_concat(alterTableFKeyConstraintList,
-												   commandFKeyConstraintList);
+		List *commandForeignKeyConstraintList =
+			GetAlterTableCommandFKeyConstraintList(command);
```

Comment:
I think it safer to skip if `commandForeignKeyConstraintList  == NIL`

Reply:
afaik `list_concat` already handles empty lists, so do you think we need extra check here ?

Reply:
Yes, it'd be cleaner